### PR TITLE
Allow context to set Kafka headers

### PR DESCRIPTION
**Pull request by AlecY**
Added additional context functions which support setting Kafka headers

### CODE REVIEW

**Comment by frairon** [comment]
Hey @AlecY, 
many thanks for the contribution! The code looks good in general, but there's one issue: the headers are sent to Kafka but they're not persisted to local disk, so neither views nor the processor itself can access them, which seems inconsistent.
So I'm wondering what would be the use case behind having this functionality?
Thanks a lot!

**Comment by AlecY** [comment]
Hey @frairon,
My use case for this functionality is using Kafka headers to send tracing information between services. I have services that read another services persist topic as an input for further processing and I'd like tracing information to be written into the persist topic so it can be read when it's processed as an input. I agree that it would be more consistent if headers where persisted to disk. However I don't require it for my use case, and I don't see a way to add header information to storage without a breaking change.

**Comment by frairon** [comment]
Hey AlecY, 
your use case makes sense and we should definitely extend the interface to support it one way or another.

We had some internal discussions how to approach this, because to be completely honest, I'm
not too happy with adding an extra version of each function just to support headers. 

We thought of two alternatives:
We could store the headers in the context prior to usage, like this:

```go
func callback(ctx goka.Context, msg interface{}){

  // setting the header, so the context will use them on every call
  // where it can be applied.
  ctx.SetHeaders(map[string]{
    "key": []byte{"value"},
    })

  // setvalue/emit with headers
  ctx.SetValue(..)
  ctx.Emit(..)

  // suppose we want some calls not to send the headers anymore
  ctx.ClearHeaders()
  // emit without headers
  ctx.Emit(..)
}

```
This approach especially makes sense if the headers are identical within one callback execution.


One other alternative (we'd prefer that :)), would be to add extendable options to the context-functions.
This means no api-change and a great deal of flexibility if there are more options to be added in future.
Something like this:

```go

type ctxOptions struct{
  emitHeaders map[string][]byte
}

type ContextOption func(*ctxOptions)

// CO-prefix to make it distinguishable from the Processor/Emitter/View-Options, but there
// are probably better names
func WithCOEmitHeaders(headers map[string][]byte) ContextOption{
  return func(opts *ctxOptions){
    opts.headers = headers
  }
}

// the context-interface would be modified like this:
type Context interface {
  // ...	
  Emit(topic Stream, key string, value interface{}, options ...ContextOption)
    
  Loopback(key string, value interface{}, options ...ContextOption)
  // modify the other functions too...
}


// how to use it in a callback
func callback(ctx context.Context, msg interface{}){
  // emit with headers
  ctx.Emit("topic", "key", "some-message", WithCOEmitHeaders(map[string][]byte{"header":[]byte{"value"}}))

  // emit without headers
  ctx.Emit("topic", "key", "some-message")
}

```

But again, those are just rough proposals, maybe there are even more elegant solutions, so we're open for discussions how we could solve this issue.

**Comment by AlecY** [comment]
Hi @frairon,

Thank you for taking the time to post such a detailed reply. I hope I've understood your comment correctly that you prefer the second alternative you posted. I have updated the pull request to use an option based approach. If I have misunderstood you please let me know and I can switch to a solution based on your first suggestion.

File: context.go [hunk 1]
```diff
@@ -170,8 +170,16 @@ type cbContext struct {
 	wg     *sync.WaitGroup
 }
 
+func WithCtxEmitHeaders(headers map[string][]byte) ContextOption{
```
Comment (frairon) [inline context.go:173]: A small description would be nice :)

File: options.go [hunk 2]
```diff
@@ -519,3 +519,15 @@ func (opt *eoptions) applyOptions(topic Stream, codec Codec, opts ...EmitterOpti
 		opt.builders.topicmgr = DefaultTopicManagerBuilder
 	}
 }
+
+type ctxOptions struct{
+	emitHeaders map[string][]byte
+}
+
+type ContextOption func(*ctxOptions)
```
Comment (frairon) [inline options.go:527]: little comment would be nice here too.

**Review by frairon — COMMENTED** [review]
Awesome, that was exactly what I meant. Just didn't want to "dictate" the preferred solution, in case you had other ideas. Just would like the two missing comments so the go linter stops complaining and it's good to go

File: options.go (hunk 2 above)
Comment (AlecY) [reply options.go:527]: I've added a comment now.

File: context.go (hunk 1 above)
Comment (AlecY) [reply context.go:173]: Oops, I've added a comment. I also noticed this should probably be in the options file and moved it.

**Review by frairon — APPROVED** [review]
looks good to me. Any objections to merge, @R053NR07?

**Review by R053NR07 — APPROVED** [review]

**Comment by R053NR07** [comment]
No, this is awesome :)

**Comment by AlecY** [comment]
Excellent, thank you for being so responsive and helpful!

**Pull request merged**